File: snobot_sim_java/java/snobot_sim_java/com/snobot/simulator/module_wrapper/DigitalSourceWrapper.java
package com.snobot.simulator.module_wrapper;

public class DigitalSourceWrapper extends ASensorWrapper
{
    public static interface StateSetterHelper
    {
        public void setState(boolean aState);
    }

    private final StateSetterHelper mSetterHelper;
    private boolean mState;

    public DigitalSourceWrapper(int index, StateSetterHelper aSetterHelper)
    {
        super("Digital Source" + index);

        mSetterHelper = aSetterHelper;
        mState = true;
    }

    public boolean get()
    {
        return mState;
    }

    public void set(boolean aState)
    {
        boolean isUpdate = mState != aState;
        mState = aState;
        if (isUpdate)
        {
            mSetterHelper.setState(mState);
        }

    }
}

File: snobot_sim/src/main/native/cpp/SnobotSim/ModuleWrapper/DigitalSourceWrapper.cpp
/*
 * DigitalSourceWrapper.cpp
 *
 *  Created on: May 3, 2017
 *      Author: PJ
 */

#include "SnobotSim/ModuleWrapper/DigitalSourceWrapper.h"

#include "SnobotSim/PortUnwrapper.h"

DigitalSourceWrapper::DigitalSourceWrapper(int aPort) :
        AModuleWrapper("Digital Source" + std::to_string(UnwrapPort(aPort))),
        mState(true)
{
}

DigitalSourceWrapper::~DigitalSourceWrapper()
{
}

bool DigitalSourceWrapper::Get()
{
    return mState;
}

void DigitalSourceWrapper::Set(bool aState)
{
    mState = aState;
}
